allennlp/modules/token_embedders/embedding.py

```diff
@@ -181,12 +181,19 @@ def extend_vocab(self,  # pylint: disable=arguments-differ
         """
         # Caveat: For allennlp v0.8.1 and below, we weren't storing vocab_namespace as an attribute,
         # knowing which is necessary at time of embedding vocab extension. So old archive models are
-        # currently unextendable unless the user used default vocab_namespace 'tokens' for it.
+        # currently unextendable.
 
         vocab_namespace = vocab_namespace or self._vocab_namespace
         if not vocab_namespace:
-            vocab_namespace = "tokens"
-            logging.warning("No vocab_namespace provided to Embedder.extend_vocab. Defaulting to 'tokens'.")
+            # It's not safe to default to 'tokens' when we aren't sure that 'tokens'
+            # need to be extended. (Without this, several tests fail.)
+            logging.warning("No vocab_namespace provided to Embedder.extend_vocab. Extension will be no-op'.")
+            return
+
+        extended_num_embeddings = extended_vocab.get_vocab_size(vocab_namespace)
+        if extended_num_embeddings <= self.num_embeddings:
```

Comment:
Shouldn't this be `==`, not `<=`?

Reply:
Yes,  if vocab and embedding are already in sync it should be `==`. I had kept `<=` as a precaution against incorrect vocab namespace. But on second thought, if user passed an incorrect vocab namespace, it's better to raise error than a silent no-op. Will change it.

Reply:
Worth explicitly raising error for `<` case? It's only possible with user explicitly passed an incorrect vocab namespace that's smaller than embedding itself.

Reply:
Okay, I have separated `==` and `<` cases, making no-op in first and raising configuration error in second.

Reply:
@matt-gardner  Correcting this revealed a subtle issue: defaulting to `tokens` and `token_characters` namepsace can be problematic, when `num_embeddings` was used instead of vocab namespace to decide embedding size. Fixed this in last commit. 

This is up for another look now.

Reply:
.. oops, didn't realize you already reviewed again!